### ISSUE TITLE
Add a universe of strict hSets

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#3397]
I think the code below should work but it fails with error
```
I'm not sure if there should be a case for the constructor reflo,
because I get stuck when trying to solve the following unification
problems (inferred index ≟ expected index):
  {_} ≟ {_}
  refl ≟ refl
  u₁ ≟ u
  u₁ ≟ u'
Possible reason why unification failed:
  Cannot eliminate reflexive equation a = a of type A because K has
  been disabled.
when checking that the pattern reflo has type PathOver B refl u u'
```
It does work when `_≡_` and `PathOver` are both `Set`-valued, or when the `--without-K` flag is removed.

```agda
{-# OPTIONS --without-K --prop #-}

data _≡_ {A : Set} (x : A) : A → Prop where
  refl : x ≡ x

data PathOver {A : Set} (B : A → Set) {a : A} : {a' : A} (p : a ≡ a') → B a → B a' → Prop where
  reflo : {u : B a} → PathOver B refl u u

PathOver-refl-from : {A : Set} {B : A → Set} {a : A} {u u' : B a}
                 → PathOver B refl u u'
                 → u ≡ u'
PathOver-refl-from reflo = refl
```

**Comment by jespercockx** [comment]
In the `Set` case, there is only one way to solve the equation `refl =?= refl`: by using the injectivity rule, which triggers higher-dimensional unification, which then discharges the equation `a =?= a`. However, in the `Prop` case, there is another way to solve it: by simple deletion, since any proof will be trivial. So this is a case where unification is actually non-confluent, and picking the wrong rule leads to a failed unification where another rule would've succeeded. 

One cheap way to fix it would be to change the order in which unification rules are tried (currently `skipIrrelevantStrategy` is tried first). However, I don't really like that the order in which we try the rules matters so much. OTOH, we probably wouldn't want to turn unification into a full-blown backtracking algorithm either.

**Comment by jespercockx** [comment]
Actually unification **is** confluent here since the injectivity rule (and hence higher-dimensional unification) don't apply for datatypes in `Prop`. So `skipIrrelevantStrategy` is really the only thing the unifier can do here.

Maybe we could just allow the `deletion` rule when the target type is in `Prop`? It seems like it could make sense, but I'll have to check to make sure.

**Comment by jespercockx** [comment]
Do you have any intuition as to why this property should hold for `Prop`-valued equality (other than that it holds for Set-valued equality --without-K)? I tried to prove it manually by doing the desugaring of pattern matching by hand, but it seems impossible because we currently don't have `J` (or even `subst`) for `Prop`-valued equality. 

I do have plans to eventually make `Prop`-valued equality more powerful to allow `subst`/`J` for it but then still this wouldn't work for an arbitrary `Set` when `--without-K` is enabled, since that would directly imply `UIP` by proof irrelevance. So we would first need a subuniverse of (strict) hSets.

**Comment by guillaumebrunerie** [comment]
No, I don’t have any more intuition for why it should hold. It was just a guess based on the fact that it works for `Set`-valued equality. But indeed, for `Set`-valued equality, one can prove it by showing that `PathOver p u u'` implies `subst p u ≡ u'` and then specialize it for `p = refl`, but it requires `subst`…

This is the only place requiring `--with-K` (or `subst`/`J` for the `Prop`-valued identity type) in the project I’m working on at the moment, so I was hoping that it isn’t actually needed.

**Comment by guillaumebrunerie** [comment]
> This is the only place requiring `--with-K` (or `subst`/`J` for the `Prop`-valued identity type) in the project I’m working on at the moment, so I was hoping that it isn’t actually needed.

Update: by being more clever on how I define quotients (using the version of `PathOver` where `p` is a relevant (Set-valued) equality instead of an irrelevant (Prop-valued) equality), then everything goes through without needing `--with-K`.
So I’m happy now, and so far everytime I thought I needed `subst` for the `Prop`-valued equality, it turned out that I just needed to juggle between the two types of equalities slightly more carefully.

**Comment by jespercockx** [comment]
That's great to hear. One trick I found is that for types with decidable equality, it is possible to 'resurrect' Prop-valued equalities and hence to implement `subst` for them:
```agda
open import Prelude

data _≐_ {a} {A : Set a} (x : A) : A → Prop a where
  refl : x ≐ x

data False : Prop where

exfalso : ∀ {a} {A : Set a} → False → A
exfalso ()

resurrect : ∀ {a} {A : Set a} {{_ : Eq A}}
          → {x y : A} → x ≐ y → x ≡ y
resurrect {x = x} {y = y} eq with x == y
... | yes path = path
... | no  neq  = exfalso (contradiction eq)
  where
    contradiction : x ≐ y → False
    contradiction refl with neq refl
    ... | ()
```

If you're feeling a bit more adventurous, you can also use rewriting to make it work for arbitrary hSets (which I think also gives a better computational behaviour on equalities between neutral terms):
```agda
{-# BUILTIN REWRITE _≡_ #-}

record IsHSet {a} (A : Set a) : Set a where
  field
    isHSet : ∀ {x y : A} (p q : x ≡ y) → p ≡ q

postulate
  subst : ∀ {a p} {A : Set a} {{_ : IsHSet A}} (P : A → Set p)
        → {x y : A} → x ≐ y → P x → P y

  subst-comp : ∀ {a p} {A : Set a} {{_ : IsHSet A}} (P : A → Set p)
             → {x : A} (p : P x) → subst P refl p ≡ p

{-# REWRITE subst-comp #-}
```

**Comment by guillaumebrunerie** [comment]
> One trick I found is that for types with decidable equality, it is possible to 'resurrect' Prop-valued equalities and hence to implement `subst` for them:

Ah, that’s nice, thanks! (although I guess that it works because the definition of "having decidable equality" uses the Set-valued equality?)

In one place I do need to transport along equalities in Nat, so for now I just proved those equalities using the Set-valued identity type.

The only other place where I need Set-valued equalities is to define quotients (using the HIT definition). Initially I had the equality constructor valued directly in the Prop-valued identity type, but it created the issue in the OP.
There are also a couple of other places where I had to box and unbox propositions, but that works very well.

Thank you for your great work on `Prop`!
